src/common/s3_adapter.cpp

```diff
@@ -223,55 +250,30 @@ bool S3Adapter::BucketExist() {
     }
 }
 
-int S3Adapter::PutObject(const Aws::String &key,
-                  const void *buffer,
-                  const int bufferSize) {
-    Aws::S3::Model::PutObjectRequest request;
-    request.SetBucket(bucketName_);
-    request.SetKey(key);
-    request.SetBody(Aws::MakeShared<Aws::StringStream>(
-        "stream", Aws::String(static_cast<const char*>(buffer), bufferSize)));
-
-    auto response = s3Client_->PutObject(request);
-    if (response.IsSuccess()) {
-            return 0;
-    } else {
-        LOG(ERROR) << "PutObject error:"
-            << bucketName_ << key
-            << response.GetError().GetExceptionName()
-            << response.GetError().GetMessage();
-        return -1;
-    }
-}
-
-int S3Adapter::PutObject(const Aws::String &key,
-                  const std::string &data) {
+int S3Adapter::PutObject(const Aws::String &key, const void *buffer,
+                         const size_t bufferSize) {
     Aws::S3::Model::PutObjectRequest request;
     request.SetBucket(bucketName_);
     request.SetKey(key);
-    std::shared_ptr<Aws::IOStream> input_data =
-                Aws::MakeShared<Aws::StringStream>("stream");
-    *input_data << data;
-    request.SetBody(input_data);
 
-    if (throttle_) {
-        throttle_->Add(false, data.size());
-    }
+    request.SetBody(Aws::MakeShared<PreallocatedIOStream>(
+        AWS_ALLOCATE_TAG, reinterpret_cast<char *>(const_cast<void *>(buffer)),
+        bufferSize));
 
     auto response = s3Client_->PutObject(request);
     if (response.IsSuccess()) {
         return 0;
     } else {
-        LOG(ERROR) << "PutObject error:"
-                << bucketName_
-                << "--"
-                << key
-                << "--"
-                << response.GetError().GetExceptionName()
-                << response.GetError().GetMessage();
+        LOG(ERROR) << "PutObject error, bucket: " << bucketName_
+                   << ", key: " << key << response.GetError().GetExceptionName()
+                   << response.GetError().GetMessage();
         return -1;
     }
 }
+
+int S3Adapter::PutObject(const Aws::String &key, const std::string &data) {
+    return PutObject(key, data.data(), data.size());
```

Comment:
Will this function cause a double free problem? Because std::string data will free it's buffer also.